### PR TITLE
Add support of TestingBot cloud service

### PR DESCRIPTION
**Pull request by OndraM**
Added full-featured support of https://testingbot.com cloud service, including publishing test results to [TestingBot API](https://testingbot.com/support/api#updatetest) (this is very similar to Sauce Labs, so most of the changes are actually refactoring of the SauceLabs specific code to make it reusable).

### CODE REVIEW

File: src-tests/ConfigHelper.php [hunk 1]
```diff
@@ -28,7 +28,7 @@ public static function getDummyConfig()
             'BROWSER_NAME' => 'firefox',
             'ENV' => 'testing',
             'SERVER_URL' => 'http://server.tld:port',
-            'CAPABILITY' => '',
+            'CAPABILITY' => '', // empty
```
Comment (jirinovak) [inline src-tests/ConfigHelper.php]: Flash of the obvious?

File: src-tests/ConfigProviderTest.php [hunk 2]
```diff
@@ -36,6 +36,14 @@ public function testShouldMakeConfigOptionsAccessibleDirectlyThroughConfigProvid
         $this->assertEquals('http://server.tld:port', ConfigProvider::getInstance()->serverUrl);
     }
 
+    public function testShouldDetectEmptyConfigOption()
+    {
+        ConfigHelper::setEnvironmentVariables($this->environmentVariables);
+
+        $this->assertFalse(empty(ConfigProvider::getInstance()->serverUrl));
+        $this->assertTrue(empty(ConfigProvider::getInstance()->capability));
```
Comment (jirinovak) [inline src-tests/ConfigProviderTest.php]: `assertEmpty`?

Comment (OndraM) [inline src-tests/ConfigProviderTest.php]: No, it won't detect if the __isset method is working properly, because PHPUnit will dereference the value by copying it into custom variable and then running empty() on it. So the test will still pass even if the _isset method is missing. You actaully really need to test directly either empty(), isset() or the __isset call...

File: src-tests/ConfigHelper.php (hunk 1 above)
Comment (OndraM) [inline src-tests/ConfigHelper.php]: Well, I want to emphasize this value is intentionally empty, as some tests depends on the emptiness. But OK, this is not so obvious :-).

File: src-tests/Publisher/AbstractCloudPublisherTestCase.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    /** @var \PHPUnit_Framework_MockObject_MockObject|AbstractTestCaseBase */
+    protected $testInstanceMock;
+
+    public function setUp()
+    {
+        $this->testInstanceMock = $this->getMockBuilder(AbstractTestCaseBase::class)
+            ->getMock();
+    }
+
+    public function testShouldDoNothingWhenPublishingTestcaseResults()
```
Comment (jirinovak) [inline src-tests/Publisher/AbstractCloudPublisherTestCase.php:25]: typo `TestCase`

File: src/Publisher/AbstractCloudPublisher.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    const CONTENT_TYPE = 'application/x-www-form-urlencoded';
+
+    public function publishResults(
+        $testCaseName,
+        $status,
+        $result = null,
+        \DateTimeInterface $startDate = null,
+        \DateTimeInterface $endDate = null
+    ) {
+        // we publish only result for each separate test using publishResult()
```
Comment (jirinovak) [inline src/Publisher/AbstractCloudPublisher.php:23]: IMHO it would be better to throw an Exception.

Because when user call this funciton nothing will happen. And now what? Did I misconfigure something?

Or I don`t get it at all :-)

Comment (OndraM) [inline src/Publisher/AbstractCloudPublisher.php:23]: This method is not called by user, it is called from the Steward (TestStatusListener) and the method it is called on all registered listeners.

**Comment by OndraM** [comment]
@jirinovak thanks for codereview!
